File: packages/features/window-management/src/services/windowService.ts
import { getCurrentWindow, getAllWindows } from '@tauri-apps/api/window';
import { WebviewWindow } from '@tauri-apps/api/webviewWindow';
import { saveWindowState, restoreStateCurrent, StateFlags } from '@tauri-apps/plugin-window-state';
import type { WindowConfig, WindowState, WindowPosition, WindowSize } from '../types';

export class WindowService {
  private static instance: WindowService;
  
  private constructor() {}
  
  static getInstance(): WindowService {
    if (!WindowService.instance) {
      WindowService.instance = new WindowService();
    }
    return WindowService.instance;
  }
  
  // Window creation and management
  async createWindow(label: string, config: WindowConfig): Promise<WebviewWindow> {
    const webview = new WebviewWindow(label, {
      url: config.url || '/',
      title: config.title,
      width: config.width,
      height: config.height,
      minWidth: config.minWidth,
      minHeight: config.minHeight,
      maxWidth: config.maxWidth,
      maxHeight: config.maxHeight,
      resizable: config.resizable,
      fullscreen: config.fullscreen,
      alwaysOnTop: config.alwaysOnTop,
      decorations: config.decorations,
      transparent: config.transparent,
      skipTaskbar: config.skipTaskbar,
      center: config.center,
      x: config.x,
      y: config.y,
    });
    return webview;
  }
  
  async closeWindow(windowId?: string): Promise<void> {
    if (windowId) {
      const window = await this.getWindow(windowId);
      return window?.close();
    }
    return getCurrentWindow().close();
  }
  
  async getWindow(windowId: string): Promise<WebviewWindow | null> {
    const windows = await getAllWindows();
    return (windows.find(w => w.label === windowId) || null) as WebviewWindow | null;
  }
  
  async getAllWindows(): Promise<WebviewWindow[]> {
    return getAllWindows() as any as WebviewWindow[];
  }
  
  async getCurrentWindow(): Promise<WebviewWindow> {
    return getCurrentWindow() as any;
  }
  
  // Window state operations
  async minimize(windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.minimize();
  }
  
  async maximize(windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.maximize();
  }
  
  async unmaximize(windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.unmaximize();
  }
  
  async toggleMaximize(windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.toggleMaximize();
  }
  
  async setFullscreen(fullscreen: boolean, windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setFullscreen(fullscreen);
  }
  
  async show(windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.show();
  }
  
  async hide(windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.hide();
  }
  
  async setFocus(windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setFocus();
  }
  
  async center(windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.center();
  }
  
  // Window properties
  async setTitle(title: string, windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setTitle(title);
  }
  
  async setAlwaysOnTop(alwaysOnTop: boolean, windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setAlwaysOnTop(alwaysOnTop);
  }
  
  async setDecorations(decorations: boolean, windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setDecorations(decorations);
  }
  
  async setSkipTaskbar(skip: boolean, windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setSkipTaskbar(skip);
  }
  
  // Window position and size
  async setPosition(position: WindowPosition, windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setPosition(new LogicalPosition(position.x, position.y));
  }
  
  async setSize(size: WindowSize, windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setSize(new LogicalSize(size.width, size.height));
  }
  
  async setMinSize(size: WindowSize | null, windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setMinSize(size ? new LogicalSize(size.width, size.height) : null);
  }
  
  async setMaxSize(size: WindowSize | null, windowId?: string): Promise<void> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    return window?.setMaxSize(size ? new LogicalSize(size.width, size.height) : null);
  }
  
  // Window state queries
  async getWindowState(windowId?: string): Promise<WindowState> {
    const window = windowId ? await this.getWindow(windowId) : getCurrentWindow();
    if (!window) {
      throw new Error('Window not found');
    }
    
    const [isMaximized, isMinimized, isFullscreen, isFocused, isVisible] = await Promise.all([
      window.isMaximized(),
      window.isMinimized(),
      window.isFullscreen(),
      window.isFocused(),
      window.isVisible(),
    ]);
    
    const position = await (window as any).position();
    const size = await (window as any).size();
    
    return {
      isMaximized,
      isMinimized,
      isFullscreen,
      isFocused,
      isVisible,
      isPinned: await window.isAlwaysOnTop(),
      position: { x: position.x, y: position.y },
      size: { width: size.width, height: size.height },
    };
  }
  
  // State persistence
  async saveWindowState(flags?: number): Promise<void> {
    return saveWindowState(flags ?? StateFlags.ALL);
  }
  
  async restoreWindowState(flags?: number): Promise<void> {
    return restoreStateCurrent(flags ?? StateFlags.ALL);
  }
}

// Import LogicalPosition and LogicalSize from Tauri
import { LogicalPosition, LogicalSize } from '@tauri-apps/api/window';

// Export singleton instance
export const windowService = WindowService.getInstance();

File: plugins/tauri-plugin-window-manager/src/window_service.rs
use crate::{Error, Result, WindowConfig, WindowInfo, WindowMessage, WindowPosition, WindowSize, WindowState, WindowTheme};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use tauri::{Emitter, Manager, Runtime, WebviewWindowBuilder};
use uuid::Uuid;

pub struct WindowManagerState<R: Runtime> {
    windows: Arc<Mutex<HashMap<String, WindowInfo>>>,
    window_order: Arc<Mutex<Vec<String>>>,
    active_window: Arc<Mutex<Option<String>>>,
    app_handle: tauri::AppHandle<R>,
}

impl<R: Runtime> WindowManagerState<R> {
    pub fn new(app_handle: tauri::AppHandle<R>) -> Self {
        Self {
            windows: Arc::new(Mutex::new(HashMap::new())),
            window_order: Arc::new(Mutex::new(Vec::new())),
            active_window: Arc::new(Mutex::new(None)),
            app_handle,
        }
    }

    pub fn create_window(&self, config: WindowConfig) -> Result<WindowInfo> {
        // Check if window already exists
        if self.app_handle.get_webview_window(&config.label).is_some() {
            return Err(Error::WindowAlreadyExists(config.label.clone()));
        }

        // Build the window
        let mut builder = WebviewWindowBuilder::new(&self.app_handle, &config.label, tauri::WebviewUrl::App(config.url.unwrap_or_else(|| "index.html".into()).into()));

        if let Some(title) = config.title {
            builder = builder.title(title);
        }
        if let Some(width) = config.width {
            if let Some(height) = config.height {
                builder = builder.inner_size(width as f64, height as f64);
            }
        }
        if let Some(x) = config.x {
            if let Some(y) = config.y {
                builder = builder.position(x as f64, y as f64);
            }
        }
        if let Some(center) = config.center {
            if center {
                builder = builder.center();
            }
        }
        if let Some(resizable) = config.resizable {
            builder = builder.resizable(resizable);
        }
        if let Some(maximized) = config.maximized {
            builder = builder.maximized(maximized);
        }
        if let Some(minimizable) = config.minimizable {
            builder = builder.minimizable(minimizable);
        }
        if let Some(closable) = config.closable {
            builder = builder.closable(closable);
        }
        if let Some(decorations) = config.decorations {
            builder = builder.decorations(decorations);
        }
        if let Some(always_on_top) = config.always_on_top {
            builder = builder.always_on_top(always_on_top);
        }
        if let Some(skip_taskbar) = config.skip_taskbar {
            builder = builder.skip_taskbar(skip_taskbar);
        }
        if let Some(visible) = config.visible {
            builder = builder.visible(visible);
        }
        // Note: transparent() method is not available in WebviewWindowBuilder
        // This feature may need to be implemented differently or removed from the config

        let window = builder.build()?;
        
        // Get window info
        let position = window.outer_position()?;
        let size = window.inner_size()?;
        let is_maximized = window.is_maximized()?;
        let is_minimized = window.is_minimized()?;
        let is_focused = window.is_focused()?;
        let is_visible = window.is_visible()?;
        
        let window_info = WindowInfo {
            id: Uuid::new_v4(),
            label: config.label.clone(),
            title: window.title()?,
            width: size.width,
            height: size.height,
            x: position.x,
            y: position.y,
            is_maximized,
            is_minimized,
            is_focused,
            is_visible,
            theme: config.theme.unwrap_or(WindowTheme::Auto),
        };

        // Store window info
        self.windows.lock().unwrap().insert(config.label.clone(), window_info.clone());
        self.window_order.lock().unwrap().push(config.label.clone());
        if config.focused.unwrap_or(true) {
            *self.active_window.lock().unwrap() = Some(config.label.clone());
        }

        Ok(window_info)
    }

    pub fn close_window(&self, label: &str) -> Result<()> {
        let window = self.app_handle.get_webview_window(label)
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))?;
        
        window.close()?;
        
        // Remove from state
        self.windows.lock().unwrap().remove(label);
        self.window_order.lock().unwrap().retain(|l| l != label);
        
        let mut active = self.active_window.lock().unwrap();
        if active.as_ref().map(|l| l == label).unwrap_or(false) {
            *active = None;
        }
        
        Ok(())
    }

    pub fn get_window(&self, label: &str) -> Result<WindowInfo> {
        self.windows.lock().unwrap()
            .get(label)
            .cloned()
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))
    }

    pub fn list_windows(&self) -> Result<Vec<WindowInfo>> {
        Ok(self.windows.lock().unwrap().values().cloned().collect())
    }

    pub fn focus_window(&self, label: &str) -> Result<()> {
        let window = self.app_handle.get_webview_window(label)
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))?;
        
        window.set_focus()?;
        *self.active_window.lock().unwrap() = Some(label.to_string());
        
        Ok(())
    }

    pub fn minimize_window(&self, label: &str) -> Result<()> {
        let window = self.app_handle.get_webview_window(label)
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))?;
        
        window.minimize()?;
        
        if let Some(info) = self.windows.lock().unwrap().get_mut(label) {
            info.is_minimized = true;
        }
        
        Ok(())
    }

    pub fn maximize_window(&self, label: &str) -> Result<()> {
        let window = self.app_handle.get_webview_window(label)
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))?;
        
        window.maximize()?;
        
        if let Some(info) = self.windows.lock().unwrap().get_mut(label) {
            info.is_maximized = true;
        }
        
        Ok(())
    }

    pub fn unmaximize_window(&self, label: &str) -> Result<()> {
        let window = self.app_handle.get_webview_window(label)
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))?;
        
        window.unmaximize()?;
        
        if let Some(info) = self.windows.lock().unwrap().get_mut(label) {
            info.is_maximized = false;
        }
        
        Ok(())
    }

    pub fn set_window_position(&self, label: &str, position: WindowPosition) -> Result<()> {
        let window = self.app_handle.get_webview_window(label)
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))?;
        
        window.set_position(tauri::Position::Physical(tauri::PhysicalPosition {
            x: position.x,
            y: position.y,
        }))?;
        
        if let Some(info) = self.windows.lock().unwrap().get_mut(label) {
            info.x = position.x;
            info.y = position.y;
        }
        
        Ok(())
    }

    pub fn set_window_size(&self, label: &str, size: WindowSize) -> Result<()> {
        let window = self.app_handle.get_webview_window(label)
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))?;
        
        window.set_size(tauri::Size::Physical(tauri::PhysicalSize {
            width: size.width,
            height: size.height,
        }))?;
        
        if let Some(info) = self.windows.lock().unwrap().get_mut(label) {
            info.width = size.width;
            info.height = size.height;
        }
        
        Ok(())
    }

    pub fn set_window_title(&self, label: &str, title: &str) -> Result<()> {
        let window = self.app_handle.get_webview_window(label)
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))?;
        
        window.set_title(title)?;
        
        if let Some(info) = self.windows.lock().unwrap().get_mut(label) {
            info.title = title.to_string();
        }
        
        Ok(())
    }

    pub fn send_message(&self, message: WindowMessage) -> Result<()> {
        let window = self.app_handle.get_webview_window(&message.to)
            .ok_or_else(|| Error::WindowNotFound(message.to.clone()))?;
        
        window.emit("window-message", &message)?;
        
        Ok(())
    }

    pub fn broadcast_message(&self, from: &str, message_type: &str, payload: serde_json::Value) -> Result<()> {
        let windows = self.windows.lock().unwrap();
        
        for label in windows.keys() {
            if label != from {
                let message = WindowMessage {
                    from: from.to_string(),
                    to: label.clone(),
                    message_type: message_type.to_string(),
                    payload: payload.clone(),
                };
                
                if let Some(window) = self.app_handle.get_webview_window(label) {
                    let _ = window.emit("window-message", &message);
                }
            }
        }
        
        Ok(())
    }

    pub fn get_window_state(&self) -> Result<WindowState> {
        let windows = self.windows.lock().unwrap().clone();
        let active_window = self.active_window.lock().unwrap().clone();
        let window_order = self.window_order.lock().unwrap().clone();
        
        Ok(WindowState {
            windows,
            active_window,
            window_order,
        })
    }

    pub fn update_window_info(&self, label: &str) -> Result<()> {
        let window = self.app_handle.get_webview_window(label)
            .ok_or_else(|| Error::WindowNotFound(label.to_string()))?;
        
        let position = window.outer_position()?;
        let size = window.inner_size()?;
        let is_maximized = window.is_maximized()?;
        let is_minimized = window.is_minimized()?;
        let is_focused = window.is_focused()?;
        let is_visible = window.is_visible()?;
        
        if let Some(info) = self.windows.lock().unwrap().get_mut(label) {
            info.width = size.width;
            info.height = size.height;
            info.x = position.x;
            info.y = position.y;
            info.is_maximized = is_maximized;
            info.is_minimized = is_minimized;
            info.is_focused = is_focused;
            info.is_visible = is_visible;
        }
        
        Ok(())
    }
}